### PR TITLE
Automatically create user work logs

### PR DESCRIPTION
**Pull request by richardguerre**
This PR adds the "Automatically create user worklogs" feature.
What it does:
1. Queries all user's activity from that has not yet been summarized into worklogs
2. Fetches all possible entities that the automatic worklogs could link to (i.e. tasks, tickets, projects, and clients).
It fetches tasks even if the user is not the developer as to allow automatic time logging of PR reviews on tasks/tickets.
3. Summarizes all the user's activities into worklogs. Each summary/worklog is about 300 seconds (5 minutes), or the length of the entire user_event if its longer than 300 seconds.
4. Groups adjacent summaries/worklogs together
5. If the first summary item is the same as the last saved summary in the db and they are 5 minutes apart, then update the last worklog in the DB.
6. For the rest, insert new user_work_logs.
7. set isSummarized to true to all user activities that have been summarized during the save.

### CODE REVIEW

File: electron/app/jobs/save-to-db.ts [hunk 1]
```diff
@@ -49,12 +49,387 @@ export type user_events_insert_input = {
...
+export type user_work_logs_insert_input = {
+    approvalStatus?: work_log_approval_status_types_enum | null;
+    billableToClient?: boolean | null;
+    clientId?: string | null;
+    clientProjectId?: number | null;
+    createdAt?: string | null;
+    deletedAt?: string | null;
+    endAt?: string | null;
+    id?: number | null;
+    meetingType?: work_log_meeting_types_enum | null;
```
Comment (ziahamza) [inline electron/app/jobs/save-to-db.ts]: great job with the types!

File: electron/app/jobs/save-to-db.ts [hunk 2]
```diff
@@ -49,12 +49,387 @@ export type user_events_insert_input = {
...
+                }
+
+                const chunks = chunkEvents(events, {
+                    disableSorting: true,
+                    chunkEvery: 300, // chunks every 300 seconds (5 minutes)
+                });
+
+                const summary = chunks.reduce<SummaryItem[]>((summary, chunk) => {
+                    const summarizedEvents = summarizeEvents(chunk, {
+                        threshold: 0.3, // any activity that takes 30% or more of the chunk is considered a possible summarizer of the whole chunk.
```
Comment (ziahamza) [inline electron/app/jobs/save-to-db.ts]: why not higher? or did this come from trial and error

File: electron/app/jobs/save-to-db.ts [hunk 3]
```diff
@@ -49,12 +49,387 @@ export type user_events_insert_input = {
...
+                        .map((event) =>
+                            mapKeywordsToEntity(possibleEntitiesData.data, [
+                                event.app,
+                                event.title,
+                            ]),
+                        )
+                        .sort((a, b) => entityPriority[a.type] - entityPriority[b.type])[0];
+
+                    if (
+                        mainEntityOfChunk.type !== 'other' &&
```
Comment (ziahamza) [inline electron/app/jobs/save-to-db.ts]: possibly a good idea to also not add "client_project" / "client"?

as I think they could lead to false positivies where it should be a ticket / task instead.

**Review by ziahamza — APPROVED** [review]
this is good as it does basic caching in the DB (with isSummarized column)

Happy to roll this out and see how it performs.

File: electron/app/jobs/save-to-db.ts (hunk 1 above)
Comment (richardguerre) [reply electron/app/jobs/save-to-db.ts]: I copied those from @abdul9411 's AddWoklogModal component, which was relay so everything was all in one file.

File: electron/app/jobs/save-to-db.ts (hunk 2 above)
Comment (richardguerre) [reply electron/app/jobs/save-to-db.ts]: no we could bump it higher, but I thought 0.3 was a good starting point as lets say the user is working on GED-123 in VS Code and localhost:3000 for 7 minutes in total and 3 minutes of something else like "learning React", those 10 minutes gets split into two 5-minute chunks. The first chunk gets summarized as being GED-123 as the user only worked on that and there is nothing to compare priority, but in the second chunk, only 2/5th of the chunk is GED-123, so for GED-123 to be considered, then the threshold has to be set below 0.4 (=2/5). Then once it is considered the algorithm compares priority of a task vs a learning activity, in which it prefers tasks.

File: electron/app/jobs/save-to-db.ts (hunk 3 above)
Comment (richardguerre) [reply electron/app/jobs/save-to-db.ts]: yeah I can do that until we have an easier way to edit worklogs within the dashboard.

**Pull request merged**